syntax = "proto3";

// Message types
message LogEntry {
    int32 term = 1;
    string msg = 2;
}

message Log {
    repeated LogEntry entries = 1;
}

// Raft service definition
service RaftNode {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse) {}
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse) {}
}

// Message types for Raft RPCs
message RequestVoteRequest {
    int32 term = 1;
    string candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
}

message AppendEntriesRequest {
    int32 term = 1;
    string leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int32 leader_commit = 6;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

